third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,876 @@
+                ExpData::Call(*id, Operation::Vector, Vec::new()).into_exp()
+            } else {
+                let loc = self.env.get_node_loc(*id);
+                self.env.diag(
+                    Severity::Bug,
+                    &loc,
+                    &format!(
+                        "Illegal use of non-empty Vector constant of length {}",
+                        x.len()
+                    ),
+                );
+                exp
+            }
+        } else {
+            exp
+        }
+    }
+}
+
+impl<'env> ExpRewriterFunctions for SimplifierRewriter<'env> {
+    fn rewrite_enter_scope<'a>(
+        &mut self,
+        id: NodeId,
+        vars: impl Iterator<Item = &'a (NodeId, Symbol)>,
+    ) {
+        self.visiting_binding.enter_scope();
+        self.values.enter_scope();
+        self.remapped_symbol.enter_scope();
+        for (_, sym) in vars {
+            self.visiting_binding.insert(*sym, id);
+            let sym_is_unsafe = self.unsafe_variables.contains(&(*sym, Some(id)));
+            let mapped_sym = {
+                let new_sym = if self.remapped_symbol.contains_key(sym) {
+                    // There is already a var named `sym` in the enclosing scope.
+                    // Rename this one.
+                    self.env.symbol_pool().make_unique(*sym)
+                } else {
+                    // There is no var named `sym` in an enclosing scope,
+                    // so we can just use it without renaming.
+                    *sym
+                };
+                if SIMPLIFIER_DEBUG {
+                    eprintln!(
+                        "adding symbol {} remapping to {} at node {}",
+                        sym.display(self.env.symbol_pool()),
+                        new_sym.display(self.env.symbol_pool()),
+                        id.as_usize()
+                    );
+                }
+                self.remapped_symbol.insert(*sym, new_sym);
+                new_sym
+            };
+            if SIMPLIFIER_DEBUG {
+                eprintln!(
+                    "removing values[{}] at node {}",
+                    mapped_sym.display(self.env.symbol_pool()),
+                    id.as_usize()
+                );
+            }
+            self.values.insert(
+                mapped_sym,
+                if sym_is_unsafe {
+                    SimpleValue::NonConstant
+                } else {
+                    SimpleValue::LocalVar(mapped_sym)
+                },
+            );
+        }
+    }
+
+    fn rewrite_exit_scope(&mut self, _id: NodeId) {
+        self.visiting_binding.exit_scope();
+        self.values.exit_scope();
+        self.remapped_symbol.exit_scope();
+    }
+
+    fn rewrite_local_var(&mut self, id: NodeId, sym: Symbol) -> Option<Exp> {
+        self.rewrite_to_recorded_value(id, sym)
+    }
+
+    fn rewrite_assign(&mut self, id: NodeId, lhs: &Pattern, rhs: &Exp) -> Option<Exp> {
+        let mut remove_vars = BTreeSet::new();
```

Comment:
It would be good to document here and in other cases what this does by examples, e.g. this appears to optimize `let (x, y) = (x, e)` to `let y = e`?